namcos10: More dumps/redumps, new decryption devices, and more I/O emulation

987123879113 opened the pull request:
This PR contains changes from various sources (Peter Wilhelmsen and Samuel Neves for their new decryption device code, Guru updating his readme with new games and PCB documentation, everything else me).

tmp95c061: Filled out A/D code to allow for A/D values to be returned properly. The implementation method was copied for consistency from tmp95c063. Also fixed ADMOD and INTE0AD registers enough to get the EXIO program to run.

ns10crypt: Samuel Neves and Peter Wilhelmsen provided decryption devices for GAHAHA Ippatsudou (gahaha), Golgo 13: Juusei no Chinkonka (g13jnc), Sekai Kaseki Hakken (sekaikh), and Pacman BALL (pacmball). Please make sure Samuel Neves and Peter Wilhelmsen get proper credit for these as without their work these games would be otherwise be completely unbootable.

namcos10 specific:
- Marked all games as MACHINE_IMPERFECT_SOUND in anticipation of complaints. There are various times where audio is not working as expected (stopping too soon, some distortion/glitching, etc). g13jnc also lacks MP3 support for now so BGMs won't play at all.
- Implemented the basics for EXIO and MGEXIO and split off the expansion I/O boards into their own file because the main driver was getting hard to follow since it has all of the classes for each ROM board type.
- Implemented the basics of the MEM(P3) MEM/IO board
- Added controls for more games that can boot now
- pacmball marked as BAD_DUMP (confirmed the ROMs are bad dumps after being able to boot it)
- Guru updated the readme in the header

Promotions:
- Golgo 13: Juusei no Chinkonka (Japan, GLT1 VER.A) promoted to working. BGM audio does not work but the game appears to be otherwise playable after the gun is calibrated in the operator menu.
- Kotoba no Puzzle Mojipittan (Japan, KPM1 Ver.A) (knpuzzle) promoted to working.
- GAHAHA Ippatsudou (World, GID2 Ver.A) (gahaha) promoted to working now that the required I/O board and controls are implemented. GAHAHA Ippatsudou 2 (Japan, GIS1 Ver.A) too.

DavidHaywood commented:
maybe the scope enable / disable on g13 would be better with PORT_TOGGLE than having to hold a key?

(the service switch with '0' on that seems to be a toggle too, as right now you have to keep it held down to remain in service mode)

angelosa commented:
imho no audio MP3 playback in Golgo 13 sounds like a scenario of "immersion ruined" that shouldn't be prematurely promoted to working.

987123879113 commented:
PORT_TOGGLE sounds a good solution for those two inputs. I'll go with that.

I guess I don't really consider the BGM *that* important when the rest of the gameplay and sound effects are working, but it's not a hill I'm willing to die on. It can always be promoted later but demoting it would probably be more of a bother. I'll put the MACHINE_NOT_WORKING flag back on.

rb6502 commented:
Missing BGM is *not* GAME_NOT_WORKING if the game's playable.  The Toaplan games were like that for a decade due to the undumped MCUs.

angelosa commented:
Arguably that wasn't a wise decision: shmups relies on audio cues and not having them doesn't make them "working" to my eyes.
But regardless in this specific case "immersion ruined" means that:
1. game is based over a known media;
2. it has recognizable BGMs from the media;
3. not having those means that _it's not faithful to the source material_, with all the implications from a premature promotion.

rb6502 commented:
Sorry, no.  That still doesn't qualify as not working.  And "immersion ruined" is normally used sarcastically to mock people like Comic Book Guy from the Simpsons.

angelosa commented:
I didn't coined the term, it's what people uses to mock us when marking stuff as working when it's not. And again: I wouldn't consider a Simpsons game "working" without Danny Elfman theme, the two things goes together as a whole for a tie-in software.

Gurudumps commented:
It's just BGM. All the game play sounds are coming from the Sony sound chip. Missing music is no big deal. Game works and is playable. Remove GAME_NOT_WORKING because that is just plain wrong. 
This sound situation is already covered. Add MACHINE_IMPERFECT_SOUND. Problem solved.

cuavas commented:
Is Golgo 13 missing voice acting as well?  Does it use audible cues for timing your shots?

cuavas commented on src/mame/namco/namcos10_exio.h:
```diff
@@ -0,0 +1,98 @@
+	required_device<h83007_device> m_maincpu;
+	required_shared_ptr<uint16_t> m_ram;
+	required_device<nvram_device> m_nvram;
+
+	int m_active_state;
```
If this needs to be saved, it should use an explicitly sized integer type.

rb6502 commented:
All of the gameplay sounds are through the PSX SPU.  MP3 is background music only, the chip likely can only play 1 sound at a time.

cuavas commented on src/mame/namco/namcos10_exio.cpp:
```diff
@@ -0,0 +1,210 @@
+DEFINE_DEVICE_TYPE(NAMCOS10_MGEXIO,    namcos10_mgexio_device,    "namcos10_mgexio", "Namco System 10 MGEXIO")
+
+// EXIO(G) has the bare minimum: CPLD, audio output jacks, gun I/O, and a card edge connector for additional I/O
+namcos10_exio_base_device::namcos10_exio_base_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock, uint8_t ident_code)
+	: device_t(mconfig, type, tag, owner, clock), m_ident_code(ident_code)
+{
+}
+
+namcos10_exio_base_device::namcos10_exio_base_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: namcos10_exio_base_device(mconfig, NAMCOS10_EXIO_BASE, tag, owner, clock, 0x32)
+{
+}
+
+////////////////////////////////////
+
+namcos10_exio_device::namcos10_exio_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: namcos10_exio_base_device(mconfig, NAMCOS10_EXIO, tag, owner, clock, 0x30),
+	m_maincpu(*this, "exio_mcu"),
+	m_ram(*this, "exio_ram")
+{
+}
```
If you’re putting the commas at the ends of the lines in initialiser lists, please put the colons introducing the initialiser lists at the end of the line as well, like:
```c++
namcos10_exio_device::namcos10_exio_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
	namcos10_exio_base_device(mconfig, NAMCOS10_EXIO, tag, owner, clock, 0x30),
	m_maincpu(*this, "exio_mcu"),
	m_ram(*this, "exio_ram")
{
}
```

cuavas commented on src/mame/namco/namcos10_exio.cpp:
```diff
@@ -0,0 +1,210 @@
+{
+}
+
+void namcos10_exio_device::device_start()
+{
+	save_item(NAME(m_is_active));
+}
+
+void namcos10_exio_device::device_reset()
+{
+}
```
Please call the base implementation of these functions even when it does nothing – we’ve had issues where the base implementation has had something added and some derived class was assuming it was empty.  It’s a good habit to get into.

cuavas commented on src/mame/namco/namcos10_exio.cpp:
```diff
@@ -0,0 +1,210 @@
+{
+	logerror("%s: exio_ctrl_w %04x\n", machine().describe_context(), data);
+
+	if (data == 3) {
+		m_maincpu->resume(SUSPEND_REASON_HALT);
+		m_active_state = 1;
+	}
```
Could you use a timer to activate the CPU and avoid the read-counting hack?  Presumably in real life it takes some time for the clock to stabilise and the CPU to reset.

987123879113 replied:
I can play around with a timer perhaps. The game I was working on had a set up that it could check to make sure it wasn't active and then wait for it to be active. If it somehow became active before it started checking the status register it'd just fail entirely so I think there's probably some expected delay there.

987123879113 replied:
I'll come back to this soon-ish (sometime within the week), wait for this before merging. This is my first time working with H8 anything so I think I just missed a lot. Going to work through the MGEXIO's init routine and see if there's something more obvious that should be the trigger.

angelosa commented:
Nope, it's background music, voice acting and likely audio cues, that's from what I gathered from last Haze's stream. It's also very miserable without all along.

cuavas commented on src/mame/namco/ns10crypt.h:
```diff
@@ -157,6 +157,51 @@ class startrgn_decrypter_device : public ns10_type2_decrypter_device
 	static const ns10_crypto_logic crypto_logic;
 };
 
+class gahaha_decrypter_device : public ns10_type2_decrypter_device
+{
+public:
+	gahaha_decrypter_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+private:
+	static uint16_t nonlinear_calc(uint64_t previous_cipherwords, uint64_t previous_plainwords, const gf2_reducer &reducer);
+	static const ns10_crypto_logic crypto_logic;
+};
+
+class g13jnc_decrypter_device : public ns10_type2_decrypter_device
+{
+public:
+	g13jnc_decrypter_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+private:
+	static uint16_t nonlinear_calc(uint64_t previous_cipherwords, uint64_t previous_plainwords, const gf2_reducer &reducer);
+	static const ns10_crypto_logic crypto_logic;
+};
```
Not really critical at the moment, but if the only difference between these devices a reference to a structure and a function, they could be reduced to a single device with the structure and function (as a lambda) supplied in machine configuration of the client.

987123879113 replied:
I'll see about writing it to work in the machine configuration. I have a feeling that would be easier to maintain too because right now you have to touch a lot of places just to add a new decrypter.

987123879113 replied:
Done. Reduced down to just `NS10_TYPE2_DECRYPTER` and `MRDRILR2_DECRYPTER` (only known Type 1 game). Tested all games with decrypters to make sure everything is still booting.

cuavas commented on src/mame/namco/namcos10.cpp:
```diff
@@ -748,6 +955,18 @@ void namcos10_state::namcos10_base(machine_config &config)
+	m_psx_remap_timer = timer_alloc(FUNC(namcos10_state::psx_remap_tick), this);
+	m_psx_remap_timer->adjust(attotime::never);
+
+	m_maincpu->space(AS_PROGRAM).install_write_tap(
+		0x1f801010, 0x1f801013,
+		"rom_configure_tap_w",
+		[this] (offs_t offset, u32 &data, u32 mem_mask) {
+			// reinstall memory map after rom_config_w finishes with update_rom_config
+			m_psx_remap_timer->adjust(attotime::zero);
+		}
+	);
```
This feels like a gross way of doing this, relying on too much knowledge of how the PSX CPU device works.

It would be best if the CPU could just get a bus error on unmapped accesses so it didn’t need to nuke everything in the address space – @galibert can the memory system do this yet, or do we still need to rely on installing bus error handlers?

The CPU could expose a notification when the when it reconfigures its mappings (see src/lib/util/notifier.h – it’s used in src/emu/emumem.h for address space mapping change notifications, and for various things by the Lua engine in src/frontend/mame/luaengine.cpp).

If you want to do it non-intrusively, you can add a change notifier on the address space (see `add_change_notifier`) and use a guard variable to ensure you don’t keep scheduling yourself reentrantly.

And that aside, you don’t need an allocated timer for this, you can just use `machine().scheduler().synchronize()` to get yourself a callback from the scheduler at the next end-of-timeslice.

@galibert do you have any more thoughts on the best way to deal with the PSX CPU nuking the input mappings?

987123879113 replied:
This code was more of a compromise because it was a hard blocker for the rest of the driver but I also didn't want to touch the PSX core code. `add_change_notifier` looks interesting though, I'll read that code a little more to see if I can apply it here.

987123879113 replied:
I tried `add_change_notifier` but it doesn't trigger anything at all when I use it on the program space of m_maincpu. Just to test I even tried it on the `m_program` stored inside the PSX core itself and even that doesn't trigger anything.

cuavas replied:
You did hold on to the subscription returned by `add_change_notifier`, I hope?  If you don’t store it somewhere, it will be immediately destructed, and you’ll be unsubscribed immediately (hence receiving no notifications of changes).

987123879113 replied:
Yeah... that was the problem. Didn't realize. It's working now and looks much cleaner than the other things. I tried it with Gahaha (which needs analog I/O so it's an easy test) and also a game that doesn't use a remapper (Medal no Tatsujin, I/O board not emulated yet so no mapping needed) and it seems to be working fine. Thanks for being patient, didn't even consider it would unsubscribe if I didn't store it.

cuavas commented on src/mame/namco/namcos10.cpp:
```diff
@@ -783,6 +1006,76 @@ TIMER_DEVICE_CALLBACK_MEMBER(namcos10_state::io_update_interrupt_callback)
+{
+	NAMCOS10_EXIO(config, m_exio, XTAL(22'118'400));
+
+	auto m_exio_cpu = subdevice<tmp95c061_device>("exio:exio_mcu");
+	m_exio_cpu->port7_write().set([this] (uint8_t data) {
+		m_exio_analog_idx = data;
+	});
```
Please expose this stuff in the EXIO device – you shouldn’t be reaching inside devices like this.

cuavas commented on src/mame/namco/namcos10.cpp:
```diff
@@ -1111,6 +1383,20 @@ void namcos10_memm_state::ns10_mrdrilr2(machine_config &config)
+
+void namcos10_memn_state::pio_dma_read(uint32_t *p_n_psxram, uint32_t n_address, int32_t n_size)
+{
+	uint16_t *ptr = reinterpret_cast<uint16_t*>(&p_n_psxram[n_address / 4]);
+	for (int32_t i = 0; i < n_size * 2; i++) {
+		ptr[i] = nand_data_r();
+	}
```
You need to use `auto ptr = little_endian_cast<uint16_t>(&p_n_psxram[n_address / 4]);` (from src/lib/util/endianness.h) to avoid endianness issues.

cuavas commented on src/mame/namco/namcos10.cpp:
```diff
@@ -1714,54 +2008,118 @@ void namcos10_memn_state::ns10_unks10md(machine_config &config)
-			dst++;
-		}
+	if (data == 1) {
+		// This should happen in the device itself but it would require calling device_reset on the MCU
+		auto pc = ((m_mcu_ram[0x7fff02 / 2] << 16) | m_mcu_ram[0x7fff00 / 2]) & 0xffffff;
+		m_memp3_mcu->set_state_int(TLCS900_PC, pc);
+		m_memp3_mcu->resume(SUSPEND_REASON_HALT);
```
You can’t call `->reset()` on the MCU?

987123879113 replied:
I thought the only way to do it was to call `device_reset()` which I remember getting told not to do before so I came up with something else to avoid it. This is cleaner and appears to work after a quick test, thanks.

cuavas commented:
Oh, one other thing, judging from the furigana above “鎮魂歌” on the Golgo 13 title screen, it’s supposed to be read “Golgo 13: Juusei no Requiem”.

Gurudumps commented:
Juusei no Chinkonka as an English title is Golgo 13 The Requiem for Gunshots.
Requiem is an English word so I doubt that's in the Japanese title.
Probably needs someone who reads and speaks Japanese fluently to confirm the title.

happppp commented:
It's about the small letters above the kanji, it indeed says レクイエム.
Put it in google translate to clear your doubt.

FoxhackDN commented:
It seems both titles are used interchangeably by Japanese players. Sometimes they use Requiem, sometimes the Kanji. I'd leave the Kanji title as the main one since that's what Bandai Namco uses (see https://bandainamco-am.co.jp/am/vg/golgo13-3/ ) and use the other as an alternate.

Gurudumps commented:
I just checked my working game for music. There's NO music within the game play. The music only plays on the cut-scenes and other places before or after game play scenes but NOT in game. Regardless MACHINE_IMPERFECT_SOUND is the way to deal with it until the MP3 decoder chip is emulated.

DopefishJustin commented:
It says レクイエム over the kanji which means the kanji are to be read Requiem and not Chinkonka, period, no ifs ands or buts.

cuavas commented on src/mame/namco/namcos10.cpp:
```diff
@@ -494,10 +654,22 @@ class namcos10_state : public driver_device
+		, m_analog(*this, "ANALOG%u", 1U)
 	{ }
 
 	void namcos10_base(machine_config &config);
 
+	void namcos10_exio(machine_config &config);
+	void namcos10_mgexio(machine_config &config);
+	void namcos10_exfinalio(machine_config &config);
+
+	void namcos10_map_exio_inner(address_map &map);
+	void namcos10_map_exio(address_map &map);
+
+	void namcos10_map_mgexio_inner(address_map &map);
+	void namcos10_map_mgexio(address_map &map);
+
 	void namcos10_map_inner(address_map &map);
 	void namcos10_map(address_map &map);
```
Can at least some of these be in the `protected` section?

cuavas commented on src/mame/namco/namcos10.cpp:
```diff
@@ -512,6 +684,7 @@ class namcos10_state : public driver_device
 	optional_device<ns10_decrypter_device> m_decrypter; // not every game has a decrypter implemented yet, so optional for now
 
 	unscramble_func m_unscrambler;
+	std::function<void()> m_psx_remapper;
```
If you turn this into a `timer_expired_delegate` then you can pass it to `synchronize` directly without an extra trampoline.

987123879113 replied:
Just to make sure I'm not misunderstanding or doing something wrong, should lambdas work with that method or should I turn the remap functions into named functions instead of using a lambda function? I gave it a go using lambdas but it always ends up giving a signature error (`lambda [](s32)->void` vs `void(s32)`).

cuavas commented on src/mame/namco/namcos10.cpp:
```diff
@@ -705,26 +887,50 @@ class namcos10_memp3_state : public namcos10_state
 
+void namcos10_state::psx_remap_io(s32 param)
+{
+	if (!m_psx_remapper)
+		return;
```
You can do the check before calling `synchronize` (or possibly even before subscribing to changes) and avoid unnecessary trips through the scheduler.

987123879113 commented:
I think I've addressed all of the comments so far aside from what to do about Golgo 13's status. 

I hope I didn't make too much of a mess of the MGEXIO code. I removed some hacky stuff I did to test games with a faked pusher motor but it's mostly working outside of the pusher motor and a divider sensor error *sometimes* showing up on sekaikh. ballpom is pretty lenient, pacmball can get into the test menu so I tested the I/O on that. Sugorotic JAPAN freezes due to an SPU issue (seems to want an SPU IRQ???) but looks good in the test menu. I don't play medal games, don't understand how to play them, and don't understand how the hardware in them so I tried to map the I/O out but it's going to require someone who understands medal game I/O to make something usable out of it.

angelosa commented:
> I think I've addressed all of the comments so far aside from what to do about Golgo 13's status.

It's not blocking for merging: what it really needs is an extensive playthrough session, which can be done on a later date.

happppp commented:
Since you renamed the Golgo game, could you rename the romset to g13jnr?

cuavas commented:
This looks OK to merge now.  There are a few trivial things I’ll clean up afterwards.

cuavas commented:
Cleanup done: 05e8f2644155317e3eb521a64ce8e0db1258afe7
